### PR TITLE
Instance retag validation check api

### PR DESCRIPTION
**Pull request by shounakmk219**
Expose an endpoint to perform safety check validation of the instance retag/ tag update operation. 
Right now in order to update the tags on an instance we use

- `/instances/{instanceName}/updateTags` 

**Problem**
This operation blindly updates the instance tags without taking into consideration the possible consequences such as table replication not being satsfied. 

**Proposal**
Provide a validation API that takes the desired user state as input and provides the user with the information if the retag is safe else give all the possible reasons for it being unsafe.

Endpoint : `/instances/updateTags/validate`
Method type : `POST`
Request body : 
```
[
  {
    "instanceName": "Server_a.c.com_20000",
    "newTags": [
      "tenantA_OFFLINE",
      "tenantA_REALTIME"
    ]
  },
  {
    "instanceName": "Server_b.c.com_20000",
    "newTags": [
      "tenantB_OFFLINE",
      "tenantB_REALTIME"
    ]
  }
]
```
Sample Response : 

```
[
  {
    "instanceName": "Broker_127.0.0.1_8000",
    "issues": [
      {
        "code": "MINIMUM_INSTANCE_UNSATISFIED",
        "message": "Tenant 'DefaultTenant' will not satisfy minimum 'broker' requirement if tag 'DefaultTenant_BROKER' is removed from broker instance 'Broker_127.0.0.1_8000'."
      },
      {
        "code": "UNRECOGNISED_TAG_TYPE",
        "message": "The tag 'tenantA' does not follow the suffix convention of either broker or server"
      }
    ],
    "isSafe": false
  },
  {
    "instanceName": "Server_127.0.0.1_7050",
    "issues": [
      {
        "code": "MINIMUM_INSTANCE_UNSATISFIED",
        "message": "Tenant 'DefaultTenant' will not satisfy minimum 'server' requirement if tag 'DefaultTenant_REALTIME' is removed from server instance 'Server_127.0.0.1_7050'."
      }
    ],
    "isSafe": false
  },
  {
    "instanceName": null,
    "issues": [
      {
        "code": "ALREADY_DEFICIENT_TENANT",
        "message": "Tenant 'DefaultTenant_OFFLINE' is low on 'server' instances by 1 even with given allocation"
      }
    ],
    "isSafe": false
  }
]
```

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 1]
```diff
@@ -4128,6 +4129,29 @@ public PeriodicTaskInvocationResponse invokeControllerPeriodicTask(String tableN
...
+   * Construct a map of all the tenants and their respective minimum server requirements.
+   * The minimum server requirement is computed by iterating over all the tables of the tenant and
+   * find the table with maximum replication.
+   * @return map of tenants and their minimum server requirements
+   */
+  public Map<String, Integer> minimumServersRequiredForTenants() {
+    Map<String, Integer> tenantMinServerMap = new HashMap<>();
+    getAllServerTenantNames().forEach(tenant -> tenantMinServerMap.put(tenant, 0));
+    for (String table : getAllTables()) {
+      TableConfig tableConfig = getTableConfig(table);
```
Comment (saurabhd336) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:4142]: Could we fetch the tableConfigs all at once instead? We may pass list<TableConfigs> to this function

Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:4142]: Makes sense to pull all the table configs at once and iterate on them directly. Not sure if passing the list would be a good option atleast for now as that puts unnecessary burden on the caller function.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 2]
```diff
@@ -4128,6 +4138,33 @@ public PeriodicTaskInvocationResponse invokeControllerPeriodicTask(String tableN
...
+   * - for SERVER tenant tag iterate over all the tables of that tenant and find the maximum table replication.
+   * - for rest of the tags just set it to 0
+   * @return map of tags and their minimum instance requirements
+   */
+  public Map<String, Integer> minimumInstancesRequiredForTags() {
+    Map<String, Integer> tagMinServerMap = new HashMap<>();
+    for (InstanceConfig instanceConfig : getAllHelixInstanceConfigs()) {
+      for (String tag : instanceConfig.getTags()) {
+        tagMinServerMap.put(tag, TagNameUtils.isBrokerTag(tag) ? 1 : 0);
+      }
```
Comment (saurabhd336) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:4154]: Does this mean for broker tags, atleast 1 broker is always needed even if no table is tagged to it?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 3]
```diff
@@ -4128,6 +4138,33 @@ public PeriodicTaskInvocationResponse invokeControllerPeriodicTask(String tableN
...
+    }
+    for (TableConfig tableConfig : getAllTableConfigs()) {
+      String serverTag = TagNameUtils.getServerTagForTenant(tableConfig.getTenantConfig().getServer(),
+          tableConfig.getTableType());
+      int maxReplication = Math.max(Objects.requireNonNullElse(tagMinServerMap.get(serverTag), 0),
+          tableConfig.getReplication());
+      tagMinServerMap.put(serverTag, maxReplication);
```
Comment (saurabhd336) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:4161]: Can be replaced with
``tagMinServerMap.put(serverTag, Math.max(tagMinServerMap.getOrDefault(serverTag, 0), tableConfig.getReplication()) )```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 4]
```diff
@@ -416,4 +422,115 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+    return tagDeficiency;
+  }
+
+  private Map<String, Integer> getUpdatedTagInstanceMap(List<InstanceTagUpdateRequest> instances) {
+    Map<String, Integer> updatedTagInstanceMap = new HashMap<>();
+    Set<String> visitedInstances = new HashSet<>();
+    instances.forEach(instance -> {
+      instance.getNewTags().forEach(tag -> {
+        Integer count = updatedTagInstanceMap.get(tag);
+        updatedTagInstanceMap.put(tag, count != null ? count + 1 : 1);
```
Comment (saurabhd336) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:521]: `updatedTagInstanceMap.put(tag, updatedTagInstanceMap.getOrDefault(tag, 0) + 1)`

Comment (saurabhd336) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:521]: same below

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 2 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java]: Now that you point out I guess there's no harm to have 0 brokers as long as it hosts no tables. Fixed it.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 3 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java]: Done

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 4 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java]: done

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 5]
```diff
@@ -416,4 +422,112 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+      Set<String> oldTags;
+      try {
+        oldTags = new HashSet<>(_pinotHelixResourceManager.getTagsForInstance(name));
+      } catch (NullPointerException exception) {
+        throw new ControllerApplicationException(LOGGER,
+            String.format("Instance %s is not a valid instance name.", name), Response.Status.PRECONDITION_FAILED);
+      }
+      Set<String> newTags = new HashSet<>(instance.getNewTags());
+      // tags removed from instance
+      for (String tag : Sets.difference(oldTags, newTags)) {
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:454]: Does this mean that we only support update the entire tag list `List<String> tags` and not support to add an extra tag to the existing tag list?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 6]
```diff
@@ -416,4 +422,112 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+
+  @POST
+  @Path("/instances/updateTags/validate")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Check if it's safe to update the tags of the given instances. If not list all the reasons.")
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Success"),
+      @ApiResponse(code = 500, message = "Internal error")
+  })
+  public List<OperationValidationResponse> instanceTagUpdateSafetyCheck(List<InstanceTagUpdateRequest> instances) {
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:434]: `instances` -> `requests`?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 7]
```diff
@@ -416,4 +422,112 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+    if (TagNameUtils.isServerTag(tag)) {
+      return "server";
+    } else if (TagNameUtils.isBrokerTag(tag)) {
+      return "broker";
+    } else {
+      return null;
+    }
+  }
+
+  private Map<String, Integer> computeTagDeficiency(List<InstanceTagUpdateRequest> instances,
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:504]: `instances` -> `requests`?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 8]
```diff
@@ -416,4 +422,112 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Check if it's safe to update the tags of the given instances. If not list all the reasons.")
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Success"),
+      @ApiResponse(code = 500, message = "Internal error")
+  })
+  public List<OperationValidationResponse> instanceTagUpdateSafetyCheck(List<InstanceTagUpdateRequest> instances) {
+    LOGGER.info("Performing safety check on tag update request received for instances: {}",
+        instances.stream().map(InstanceTagUpdateRequest::getInstanceName).collect(Collectors.toList()));
+    Map<String, Integer> tagMinServerMap = _pinotHelixResourceManager.minimumInstancesRequiredForTags();
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:437]: Why don't we compute minInstances only for tags that show up in the request?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 7 above)
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:504]: `computeTagDeficiency()` and `getUpdatedTagInstanceMap()` are hard to follow the logic due to a lot of stream is being used. Can you add the documentation on what we are trying to compute here? Also, can we try to review the logic and see if we can simplify some parts (if possible)?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 6 above)
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:434]: It would be great if we can provide the documentation on what are we validating here.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java [hunk 9]
```diff
@@ -416,4 +422,112 @@ public List<OperationValidationResponse> instanceDropSafetyCheck(
...
+        if (tagType == null && (name.startsWith(CommonConstants.Helix.PREFIX_OF_BROKER_INSTANCE)
+            || name.startsWith(CommonConstants.Helix.PREFIX_OF_SERVER_INSTANCE))) {
+          responseMap.get(name).add(new OperationValidationResponse.ErrorWrapper(
+              OperationValidationResponse.ErrorCode.UNRECOGNISED_TAG_TYPE, tag));
+          continue;
+        }
+        Integer deficiency = tagDeficiency.get(tag);
+        if (deficiency != null && deficiency > 0) {
+          tenantIssues.add(new OperationValidationResponse.ErrorWrapper(
+              OperationValidationResponse.ErrorCode.ALREADY_DEFICIENT_TENANT, TagNameUtils.getTenantFromTag(tag),
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:476]: What's the difference between `ALREADY_DEFICIENT_TENANT` vs `MINIMUM_INSTANCE_UNSATISFIED`? 

Aren't they essentially the same error?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 5 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:454]: That's correct. Its mostly aimed to act as a prerequisite to the `/instances/{instanceName}/updateTags` endpoint so that user can validate the operation safety before performing the actual updateTags operation.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 6 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:434]: will change it

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 7 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:504]: will change it

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 8 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:437]: Just the tags in request won't do, we also need to fetch the existing/old tags first too for minInstances. We can pass a list of tags to get required minInstances but anyway we are forced to parse all the table configs hence didn’t complicate things for now. Let me know if accepting the tags list makes more sense and I will be happy to do it that way.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 7 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:504]: Sure, will add some doc and comments as well

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 6 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:434]: Will do

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 9 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java:476]: `MINIMUM_INSTANCE_UNSATISFIED ` -> The tenant currently satisfies minimum instances requirement, but the validation request removes few instances from it due to which the tenant ends up with instance deficiency.
`ALREADY_DEFICIENT_TENANT` -> The tenant is already deficient on instances, and even though the validation request adds few instances to it, its still deficient.

I have made this segregation so that user is aware about the existing deficiencies and not wonder why the validation is failing even though we are adding instances to a tenant.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java (hunk 7 above)
Comment (shounakmk219) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceRestletResource.java]: Added the docs and some comments, also made few changes to these utility methods to make the code flow more readable

**Review by snleee — APPROVED** [review]
LGTM. Thank you for addressing all the comments!

**Comment by Jackie-Jiang** [comment]
@shounakmk219 The new added test is flaky, can you please take a look?

```
Error:  Failures: 
Error:    PinotInstanceRestletResourceTest.instanceRetagHappyPathTest:211->getCurrentInstanceTagsMap:282->getInstanceTags:291->ControllerTest.sendGetRequest:770->ControllerTest.sendGetRequest:780 » IO
Error:    PinotInstanceRestletResourceTest.instanceRetagServerDeficiencyTest:237->getCurrentInstanceTagsMap:282->getInstanceTags:291->ControllerTest.sendGetRequest:770->ControllerTest.sendGetRequest:780 » IO
Error:    PinotInstanceRestletResourceTest.testInstanceListingAndCreation:74->checkNumInstances:163 expected [8] but found [9]
```

https://github.com/apache/pinot/actions/runs/5959747075/job/16165968277
